README.md
# Galgame Spider
#### 不知道这是拿来干啥的（逃
---


环境配置：

- Scrapy:
``` bash
pip install Scrapy
```

- lxml: 大多数发行版Linux自带lxml，如果缺失请查看<http://lxml.de/installation.html>

- OpenSSL 除了Windows之外的系统基本已经提供


启动：
在galgame_spider目录下：
``` bash
scrapy crawl MGS
```


JSON存在 `/game_list.json`中

能方便看懂的结果在 `/game_list`中

~~由于Scrapy的Request是异步的所以顺序全乱了(~~



changelog
2018-02-26 10:45:19 共爬取1条, 耗时297秒
2018-02-25 23:18:05 共爬取1条, 耗时365秒
2018-02-24 00:06:01 共爬取2条, 耗时359秒
2018-02-23 01:46:33 共爬取2条, 耗时6136秒
2018-02-20 01:48:18 共爬取2条, 耗时5996秒
2018-02-19 00:07:43 共爬取2条, 耗时461秒
2018-02-18 01:57:08 共爬取2条, 耗时6585秒
2018-02-16 03:25:54 共爬取2条, 耗时12032秒
2018-02-15 01:32:22 共爬取2条, 耗时5527秒
2018-02-12 01:32:51 共爬取554条, 耗时4803秒
2018-02-10 01:29:32 共爬取574条, 耗时4929秒
2018-02-04 01:35:30 共爬取574条, 耗时5174秒
2018-02-03 22:57:24 共爬取574条, 耗时168960秒
2018-02-01 01:41:48 共爬取574条, 耗时5966秒
2018-01-31 01:43:20 共爬取574条, 耗时5998秒
2018-01-30 01:55:04 共爬取574条, 耗时6582秒
2018-01-29 11:05:17 共爬取574条, 耗时238秒
2018-01-29 00:39:19 共爬取574条, 耗时2334秒
2018-01-28 12:51:45 共爬取574条, 耗时267秒
2018-01-27 10:04:37 共爬取574条, 耗时245秒
2018-01-26 10:37:47 共爬取574条, 耗时226秒
2018-01-26 01:55:54 共爬取574条, 耗时6573秒
2018-01-25 09:34:43 共爬取574条, 耗时34102秒
2018-01-25 09:34:41 共爬取446条, 耗时177秒
2018-01-24 01:04:58 共爬取574条, 耗时3762秒
2018-01-23 10:42:37 共爬取574条, 耗时224秒
2018-01-22 09:57:23 共爬取574条, 耗时223秒
2018-01-22 01:38:53 共爬取574条, 耗时5371秒
2018-01-21 10:14:36 共爬取574条, 耗时221秒
2018-01-21 01:46:53 共爬取574条, 耗时5970秒
2018-01-20 10:27:06 共爬取574条, 耗时225秒
2018-01-19 11:46:57 共爬取174条, 耗时123秒
2018-01-18 00:04:06 共爬取446条, 耗时245秒
2018-01-18 00:00:15 共爬取574条, 耗时344秒
2018-01-17 00:09:22 共爬取574条, 耗时246秒
2018-01-16 12:47:11 共爬取574条, 耗时224秒
2018-01-15 13:08:24 共爬取574条, 耗时224秒
2018-01-14 01:05:25 共爬取574条, 耗时224秒
2018-01-13 08:54:06 共爬取574条, 耗时1119秒
2018-01-12 14:56:51 共爬取574条, 耗时224秒
2018-01-11 00:04:20 共爬取574条, 耗时258秒
2018-01-10 12:43:27 共爬取574条, 耗时227秒
2018-01-09 00:04:19 共爬取574条, 耗时258秒
2018-01-08 00:23:06 共爬取574条, 耗时1297秒
2018-01-07 00:05:47 共爬取574条, 耗时345秒
2018-01-06 00:04:53 共爬取574条, 耗时291秒
2018-01-05 00:04:28 共爬取574条, 耗时266秒
2018-01-04 00:23:08 共爬取574条, 耗时1301秒
2018-01-03 12:09:14 共爬取574条, 耗时223秒
2018-01-02 11:29:10 共爬取574条, 耗时222秒
2018-01-01 00:04:59 共爬取574条, 耗时255秒
2017-12-31 11:24:21 共爬取574条, 耗时223秒
2017-12-30 13:36:38 共爬取574条, 耗时234秒
2017-12-29 09:32:16 共爬取574条, 耗时223秒
2017-12-28 00:31:14 共爬取574条, 耗时1873秒
2017-12-27 00:12:26 共爬取574条, 耗时222秒
2017-12-26 11:30:19 共爬取574条, 耗时223秒
2017-12-25 02:36:31 共爬取574条, 耗时225秒
2017-12-24 00:21:40 共爬取574条, 耗时226秒
2017-12-23 00:29:39 共爬取574条, 耗时226秒
2017-12-22 00:04:32 共爬取574条, 耗时271秒
2017-12-21 00:27:36 共爬取574条, 耗时223秒
2017-12-20 00:19:37 共爬取574条, 耗时1122秒
2017-12-19 00:22:52 共爬取574条, 耗时1311秒
2017-12-18 00:50:34 共爬取574条, 耗时223秒
2017-12-17 10:53:15 共爬取574条, 耗时39170秒
2017-12-16 00:05:10 共爬取574条, 耗时309秒
2017-12-15 00:32:04 共爬取574条, 耗时223秒
2017-12-14 00:36:48 共爬取574条, 耗时2206秒
2017-12-13 00:04:47 共爬取574条, 耗时286秒
2017-12-12 00:04:19 共爬取574条, 耗时257秒
2017-12-11 00:04:16 共爬取574条, 耗时254秒
2017-12-10 00:04:45 共爬取574条, 耗时283秒
2017-12-09 00:04:43 共爬取574条, 耗时281秒
2017-12-08 00:04:34 共爬取574条, 耗时272秒
2017-12-07 00:04:37 共爬取574条, 耗时275秒
2017-12-06 00:04:16 共爬取574条, 耗时253秒
2017-12-05 00:04:39 共爬取574条, 耗时277秒
2017-12-04 00:04:39 共爬取574条, 耗时277秒
2017-12-03 00:05:08 共爬取574条, 耗时306秒
2017-12-02 00:04:43 共爬取574条, 耗时281秒
2017-12-01 01:26:00 共爬取574条, 耗时260秒


galgame_spider/pipelines.py
# -*- coding: utf-8 -*-
import codecs
import json
import time

class GalgameSpiderPipeline(object):
    def __init__(self):
        self.game_list_json = codecs.open('game_list.json','w+',encoding='utf-8')
        self.game_list = codecs.open('game_list','w+',encoding='utf-8')
        self.change_log = codecs.open('changelog', 'r+', encoding='utf-8')
        self.data = []

    def process_item(self, item, spider):
        if item != [] and item['addr'] != []:
            item['addr'] = item['addr'][0].lstrip("/").rstrip("'")
            item['title'] = item['title'][0]
            item['pwd'] = item['pwd'][0].rstrip(u"：") + ":" + item['pwd'][1]
            item['date'] = item['date'].lstrip()
            line = json.dumps(dict(item), ensure_ascii=False) + "\r\n"
            self.data.append(item)
            self.game_list_json.write(line)
        return '[INFO]FINISH'

    def close_spider(self, spider):
        for i in self.data:
            addr = i['addr']
            title = i['title']
            pwd = i['pwd']
            date = i['date']
            self.game_list.write(u"标题:"+title+"\r\n")
            self.game_list.write(u"日期:" + date + "\r\n")
            self.game_list.write(u"链接:"+addr+"\r\n")
            self.game_list.write(pwd+"\r\n")
            self.game_list.write("\r\n")
        time_string = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
        time_spend = int(round(time.time() - spider.start_time))
        old_log = self.change_log.read()
        self.change_log.seek(0)
        self.change_log.write(u"%s 共爬取%d条, 耗时%d秒\r\n" % (time_string, len(self.data), time_spend))
        self.change_log.write(old_log)
        self.game_list_json.close()
        self.game_list.close()
        self.change_log.close()


galgame_spider/spiders/mygalgame_spider.py
# -*- coding:utf-8 -*-
import scrapy
from scrapy.selector import Selector
from scrapy import log
import time
from galgame_spider.items import GalgameSpiderItem

class MyGalgameSpider(scrapy.Spider):
    start_time = time.time()
    name = "MGS"
    start_urls = ["http://www.mygalgame.com"]
    page_number = 72
    for i in range(2, page_number+1):
        start_urls.append("http://www.mygalgame.com/page/"+str(i)+"/")

    def parse(self, response):
        for href in response.xpath('//div[@class = "title-article"]/center/h4/a/@href'):
            url = response.urljoin(href.extract())
            yield scrapy.Request(url, callback = self.parse_dir_contents)

    def parse_dir_contents(self,response):
        item = GalgameSpiderItem()
        item['pwd'] = response.xpath('//div[@class = "panel panel-primary"]/div[@class = "panel-footer"]/b//span/text()')[0:2].extract()
        item['addr'] = response.xpath('//div[@class = "panel panel-primary"]/div[@class = "panel-body"]/a/button/@onclick').re('\\/pan\\.baidu\\.com\\/s\\/.*[^.)]')
        item['title'] = response.xpath('//div[@class = "title-article"]/h4/a/text()').extract()
        item['date'] = response.xpath('//span[@class = "label label-zan"]/text()')[0].extract()
        yield item

galgame_spider/settings.py
# -*- coding: utf-8 -*-

# Scrapy settings for galgame_spider project
#
# For simplicity, this file contains only settings considered important or
# commonly used. You can find more settings consulting the documentation:
#
#     http://doc.scrapy.org/en/latest/topics/settings.html
#     http://scrapy.readthedocs.org/en/latest/topics/downloader-middleware.html
#     http://scrapy.readthedocs.org/en/latest/topics/spider-middleware.html

BOT_NAME = 'galgame_spider'

SPIDER_MODULES = ['galgame_spider.spiders']
NEWSPIDER_MODULE = 'galgame_spider.spiders'

USER_AGENT = 'MOZilla/5.0(x11; Linux x86_64; rv:7.0.1)Gecko/20100101 Firefox/7.7'
DOWNLOAD_DELAY = 0.25
# Crawl responsibly by identifying yourself (and your website) on the user-agent
#USER_AGENT = 'galgame_spider (+http://www.yourdomain.com)'

# Obey robots.txt rules
ROBOTSTXT_OBEY = True

# Configure maximum concurrent requests performed by Scrapy (default: 16)
#CONCURRENT_REQUESTS = 32

# Configure a delay for requests for the same website (default: 0)
# See http://scrapy.readthedocs.org/en/latest/topics/settings.html#download-delay
# See also autothrottle settings and docs
#DOWNLOAD_DELAY = 3
# The download delay setting will honor only one of:
#CONCURRENT_REQUESTS_PER_DOMAIN = 16
#CONCURRENT_REQUESTS_PER_IP = 16

# Disable cookies (enabled by default)
#COOKIES_ENABLED = False

# Disable Telnet Console (enabled by default)
#TELNETCONSOLE_ENABLED = False

# Override the default request headers:
#DEFAULT_REQUEST_HEADERS = {
#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
#   'Accept-Language': 'en',
#}

# Enable or disable spider middlewares
# See http://scrapy.readthedocs.org/en/latest/topics/spider-middleware.html
#SPIDER_MIDDLEWARES = {
#    'galgame_spider.middlewares.MyCustomSpiderMiddleware': 543,
#}

# Enable or disable downloader middlewares
# See http://scrapy.readthedocs.org/en/latest/topics/downloader-middleware.html
#DOWNLOADER_MIDDLEWARES = {
#    'galgame_spider.middlewares.MyCustomDownloaderMiddleware': 543,
#}

# Enable or disable extensions
# See http://scrapy.readthedocs.org/en/latest/topics/extensions.html
#EXTENSIONS = {
#    'scrapy.extensions.telnet.TelnetConsole': None,
#}

# Configure item pipelines
# See http://scrapy.readthedocs.org/en/latest/topics/item-pipeline.html
ITEM_PIPELINES = {
    'galgame_spider.pipelines.GalgameSpiderPipeline': 300,
}
FIELDS_TO_EXPORT = [
    'addr',
    'title',
    'pwd',
    'date'
]
# Enable and configure the AutoThrottle extension (disabled by default)
# See http://doc.scrapy.org/en/latest/topics/autothrottle.html
#AUTOTHROTTLE_ENABLED = True
# The initial download delay
#AUTOTHROTTLE_START_DELAY = 5
# The maximum download delay to be set in case of high latencies
#AUTOTHROTTLE_MAX_DELAY = 60
# The average number of requests Scrapy should be sending in parallel to
# each remote server
#AUTOTHROTTLE_TARGET_CONCURRENCY = 1.0
# Enable showing throttling stats for every response received:
#AUTOTHROTTLE_DEBUG = False

# Enable and configure HTTP caching (disabled by default)
# See http://scrapy.readthedocs.org/en/latest/topics/downloader-middleware.html#httpcache-middleware-settings
#HTTPCACHE_ENABLED = True
#HTTPCACHE_EXPIRATION_SECS = 0
#HTTPCACHE_DIR = 'httpcache'
#HTTPCACHE_IGNORE_HTTP_CODES = []
#HTTPCACHE_STORAGE = 'scrapy.extensions.httpcache.FilesystemCacheStorage'
